// st/test/basic.js
var path = require('path')
var fs = require('fs')
var zlib = require('zlib')
var http = require('http')
var server
var st = require('../st.js')
var request = require('request')
var tap = require('tap')
var test = tap.test
var port = process.env.PORT || 1337
var util = require('util')

var opts = util._extend({
  autoindex: true,
  path: path.dirname(__dirname),
  url: '/test'
}, global.options || {})

var mount = st(opts)
exports.mount = mount
exports.req = req

function req (url, headers, cb) {
  if (typeof headers === 'function') cb = headers, headers = {}
  request({ encoding: null,
            url: 'http://localhost:' + port + url,
            headers: headers }, cb)
}

test('setup', function (t) {
  server = http.createServer(function (req, res) {
    if (!mount(req, res)) {
      res.statusCode = 404
      return res.end('Not a match: ' + req.url)
    }
  }).listen(port, function () {
    t.pass('listening')
    t.end()
  })
})

tap.tearDown(function() {
  server.close()
})


var stEtag
var stExpect = f
